scipy/special/special/config.h

```diff
@@ -53,6 +53,33 @@
 #define M_SQRT1_2 0.707106781186547524401
 #endif
 
+namespace std {
+
+inline constexpr size_t dynamic_extent = numeric_limits<size_t>::max();
+
+// TODO: Do we want this?
+template <typename T, size_t Extent = dynamic_extent>
+class span {
```

Comment:
I'm not a C++ expert (so this could be noise) but could this be confused with [std::span](https://en.cppreference.com/w/cpp/container/span)?

Reply:
that was introduced in c++20, we need to support c++17 at the minute

Reply:
I knew that we support 17, but, at some point in the future we'll want to move to 20, right?

Reply:
> I knew that we support 17, but, at some point in the future we'll want to move to 20, right?

That'll probably be a while yet. C++20 was a _huge_ release and even the latest compilers still aren't fully supporting that standard yet.

Here's a pretty recent [comment](https://discourse.llvm.org/t/rfc-clang-17-0-6-would-be-minimum-version-to-build-llvm-in-c-20/75345/8?u=h-vetinari) from someone at google involved in trying to upgrade their standard version to C++20, and how that's anything but a simple switch:
> There were multiple serious issues for consteval, concepts, and other new features. For Google’s internal codebase, these caused problems just trying to switch from -std=c++17 to -std=c++20 – including silent miscompiles. Our codebase didn’t even use C++20 features, we were just switching the compilation mode (which introduced usage of C++20 features via the stdlib headers.)
> 
> For reasonable C++20 compiler support, I’ve been making a recommendation of the following minimum compiler versions:
> 
>  * Clang 17
>  * GCC 12
>  * MSVC 17.2
>
> …and don’t use coroutines or modules yet, both of which are not yet fully-baked.

While the comment mostly refers to clang, it hopefully does illustrate the mismatch of the expectation that people will upgrade to C++20 and beyond quickly. In particular, even if we forbid coroutines & modules (and stdlib features that aren't universally implemented yet) as a matter of policy, I don't think we're at a point where we'd be able to set those lower bounds on our compiler support. I've been meaning to [update](https://github.com/scipy/scipy/issues/19815) the toolchain roadmap, but in any case, don't count on C++20 for now.

Reply:
The idea of `config.h` is that it defines a consistent set of things that will work on all targeted platforms, including CUDA and puts them in the `std` namespace. If this is API compatible with `std::span`, then we just need to guard it to only define if the symbol doesn’t already exist. The library we’re developing is intended for use beyond SciPy though, so we can’t just assume C++17 because that’s the currently supported version. I think we should assume at least C++17.

Reply:
I also share @Kai-Striega's concern. We are getting too close to bad parts of C++ in my opinion with custom classes for memory wrappers. And this will add more code rot. 

Unlike Fortran, C++ is not good for longterm archive. So to yhe question in the todo comment I'd say probably not.

Reply:
I think @steppi explained it well. Essentially, for each gufunc array argument, we need to pass around the pointer, shape, and strides. We can't get away from that. Either we do something custom or we do something inline with where the standard is going.